packages/js/components/src/experimental-select-control/async-select-control.tsx

```diff
@@ -0,0 +1,168 @@
+/**
+ * External dependencies
+ */
+import { Spinner } from '@wordpress/components';
+import {
+	createElement,
+	useState,
+	useMemo,
+	useCallback,
+	useEffect,
+} from '@wordpress/element';
+import { debounce } from 'lodash';
+import { ReactElement, Component } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import { SelectControlProps, SelectControl } from './select-control';
+import { Menu } from './menu';
+import { MenuItem } from './menu-item';
+import {
+	defaultGetFilteredItems,
+	defaultGetItemLabel,
+	defaultGetItemValue,
+} from './utils';
+import { ChildrenProps } from './types';
+
+type AsyncSelectControlProps< ItemType > = Omit<
+	SelectControlProps< ItemType >,
+	'onInputChange' | 'items' | 'children'
+> & {
+	onSearch: ( value: string | undefined ) => Promise< ItemType[] >;
+	items?: ItemType[];
+	triggerInitialFetch?: boolean;
+	pageSize?: number;
+	total?: number;
+	children: (
+		props: ChildrenProps< ItemType > & {
+			isFetching: boolean;
+		}
+	) => ReactElement | Component;
+};
+
+export function AsyncSelectControl< ItemType >( {
+	onSearch,
+	getItemLabel = defaultGetItemLabel,
+	getItemValue = defaultGetItemValue,
+	items = [],
+	triggerInitialFetch = true,
+	pageSize,
+	total,
+	children = ( {
+		items: renderItems,
+		highlightedIndex,
+		getItemProps,
+		getMenuProps,
+		isOpen,
+		isFetching,
+	} ) => {
+		return (
+			<Menu isOpen={ isOpen } getMenuProps={ getMenuProps }>
+				{ isFetching ? (
+					<Spinner />
```

Comment:
Spinner feels a bit odd in a dropdown like this. Maybe if it had some text with it as well it would be better.